torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1115,57 +1162,71 @@ void ProcessGroupNCCL::abort(c10::optional<std::string> abortReason) {
   std::lock_guard<std::mutex> lock(mutex_);
   abortCommsFromMap(devNCCLCommMap_, abortReason);
   abortCommsFromMap(inInitializationCommMap_, abortReason);
+  return true;
 }
 
 void ProcessGroupNCCL::shutdown() {
   // Don't join threads here since the purpose of this method is to abort all
   // communicators and signal the threads to exit. Joining on the threads could
   // potentially block and hence avoid it in this method.
   terminateProcessGroup_.store(true);
+  workMetaListCV_.notify_one();
```

Comment:
so the intent here is to get the watchdog loop to wake up and shut itself down, right?

notifying it is good, but doesn't gaurantee it exits before our async abort kicks in.   its still a race whether ongoing works get updated by watchdog thread and observed as completed due to the abort right?  (do we care?)

Reply:
Right, only after watchdog thread joined the destructor thread can we guarantee the exits. I think the watchdog only care the collectives before the shutdown call, but watchdog would still finish its while loop for already enqueued collectives (Though in normal cases, all collectives should complete before calling destroy_process_group). Even if there are non 0 left collectives, As we discussed, watchdog can 'record' when termination happens both in log and in flight recorder, which would be a separate PR